src/initlog.1
.TH initlog 8 "Sun Jan 24 1999"
.SH NAME
initlog \- log messages and events to the system logger
.SH SYNOPSIS
.B initlog
[\-cefnpqrs] [\-\-cmd=ARG] [\-\-event=ARG] [\-\-facility=ARG]
[\-\-name=ARG] [\-\-priority=ARG] [\-\-run=ARG] [\-\-string=ARG]
.SH DESCRIPTION
\fBinitlog\fR logs messages and events to the system logger.
It is mainly designed for use in init scripts. initlog
reads a configuration file
.I /etc/initlog.conf
by default, to determine its settings. Any line preceded with a 
.I #
is a comment, and the following configuration directives
are recognized:
.TP
.I facility <logging facility>
Sets the default logging facility
.TP
.I priority <logging priority>
Sets the default logging priority
.TP
.I ignore <regexp>
Messages that match the regular expression will not be logged.
.TP
initlog behavior can also be configured by command-line options.

\fBNote that initlog is deprecated and will be removed in a future
release.\fr

.SS OPTIONS
.TP
.I "\-c, \-\-cmd=[program]"
Execute the specified program, logging anything output to
stdout or stderr.
.TP
.I "\-e, \-\-event=[number]"
Logs that the specified event happened. Used in conjunction
with \fB\-\-name\fR. Currently specified events are:
.PD 0
.RS 8n
.TP 3n
.B 1
the action completed successfully
.TP
.B 2
the action failed
.TP
.B 3
the action was cancelled at user request
.TP
.B 4
the action failed due to the failure of a dependent action
.RE
.PD
.TP
.I "\-f, \-\-facility=[facility]"
Log at the specified syslog facility. The default
is \fBdaemon\fR (see syslog(3)).
.TP
.I "\-n, \-\-name=[string]"
Log the event under the specified string, such as
"inetd".
.TP
.I "\-p, \-\-priority=[priority]"
Log at the specified syslog priority. The default
is \fBnotice\fR (see syslog(3)).
.TP
.I "\-q"
Do not print the program's output, unless it exits
with a non-zero exit code.
.TP
.I "\-r, \-\-run=[program]"
Execute the specified program, with an open file
descriptor so that the program can pass back
commands to initlog.
.TP
.I "\-s, \-\-string=[string]"
Log the specified string to the logger.
.TP
.I "\-\-conf=[file]"
Specifies an alternate configuration file.
.SH FILES
.I /etc/initlog.conf
.SH "SEE ALSO"
syslog(3), logger(1)


ChangeLog
commit f8c758709c2dea6651afcc75fa7cceb3b7ff7afc
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 18:17:22 2010 +0200

    version 8.45.32

 initscripts.spec |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 967b64ce3f9754f7de902ed290c813634c182803
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 18:16:21 2010 +0200

    52a7a1f115bb87323620b73f5dde21e2c5e7d9a7

 ChangeLog                        |    7 ++++---
 udev/rules.d/54-cu3088-fix.rules |   10 ++++++++++
 2 files changed, 14 insertions(+), 3 deletions(-)

commit c014f9c6c09b50302e89f9df21077343b6c363b8
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 17:34:33 2010 +0200

    initrscripts.spec: sync with cvs repo

 initscripts.spec |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 923ef28cc4d4654e58033c81cd44ce9e1e4734b8
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 17:30:23 2010 +0200

    version 8.45.31

 ChangeLog        | 1016 ++++++++++++++++++++++++++++++++++--------------------
 initscripts.spec |    4 +-
 2 files changed, 636 insertions(+), 384 deletions(-)

commit 52a7a1f115bb87323620b73f5dde21e2c5e7d9a7
Author: Phil Knirsch <pknirsch@redhat.com>
Date:   Mon Sep 13 16:29:45 2010 +0200

    s390: Added required 54-cu3088-fix.rules udev rules file (#475721)

 initscripts.spec |    6 ++++++
 1 files changed, 6 insertions(+), 0 deletions(-)

commit b74c943aa01f6a8facc36246dc20ccebd9eb690c
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:41:21 2008 -1000

    Expand get_hostname() function to support IPv4 and IPv6.
    
    Make sure that get_hostname() if the user specifies either address
    family.

 src/ipcalc.c |   36 +++++++++++++++++++++++++++++-------
 1 files changed, 29 insertions(+), 7 deletions(-)

commit f085cdfdc0d67fe240c23e75d14dce6534d36c8a
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:30:38 2008 -1000

    Remove unnecessary tab.

 src/ipcalc.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit e921ac88353159dee451f7c0094fca95a9b2bb56
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:29:21 2008 -1000

    Correct address family option handling in ipcalc.
    
    Make sure the default is IPv4.  If the check option is specified,
    make sure we just validate the address and return 1 or 0.

 src/ipcalc.c |   83 ++++++++++++++++++++++++++++-----------------------------
 1 files changed, 41 insertions(+), 42 deletions(-)

commit 9253ed4bdddd635706a0a741259cbaa8a21d8b89
Author: David Cantrell <dcantrell@redhat.com>
Date:   Fri Sep 26 16:26:27 2008 -1000

    Compile ipcalc with _GNU_SOURCE defined (for asprintf).

 src/Makefile |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 192392a47594d8141aaf14a0a705bc421899adb0
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 16:20:36 2010 +0200

    remove deprecation comments for raw devices and fix path

 initscripts.spec       |    6 ++++++
 rc.d/init.d/rawdevices |    6 ++----
 sysconfig/rawdevices   |    3 ---
 3 files changed, 8 insertions(+), 7 deletions(-)

commit 6c0c324a6c0c031ed4604c6a020829503e8d086f
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 15:59:40 2010 +0200

    version 8.45.30

 initscripts.spec |   28 +++++++++++++++++++++++++++-
 1 files changed, 27 insertions(+), 1 deletions(-)

commit 8f65b61fc53d245c7865bfa0ab879337d9c6ff31
Author: Harald Hoyer <harald@redhat.com>
Date:   Mon Sep 13 15:48:34 2010 +0200

    rc.sysinit: also remove /var/run/dovecot/*
    
    https://bugzilla.redhat.com/show_bug.cgi?id=492994

 rc.d/rc.sysinit |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 85f809ad1d6f47d4fe73a07fa358fd917bc94b87
Author: Bill Nottingham <notting@redhat.com>
Date:   Wed Sep 2 16:28:02 2009 -0400

    Doucment DHCP_HOSTNAME (#498052)

 sysconfig.txt |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

commit 52dfb70cd860e4d0fd51e37e30a081fd197f1b74
Author: Bill Nottingham <notting@redhat.com>
Date:   Fri Jul 20 20:54:47 2007 +0000

    move quotacheck to after filesystem mounting so it actually works (#249003, <tometzky@batory.org.pl>)
    also, nuke the convertquota stuff that has been obsolete (and not working) for quite some time

 rc.d/rc.sysinit |   25 +++++--------------------
 1 files changed, 5 insertions(+), 20 deletions(-)

commit 873660981a5a42a6de411e04958a4f204704efe1
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Dec 15 10:59:44 2009 -0500

    Silence error. (#516569)

 sysconfig/network-scripts/network-functions |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

commit 3ae2380e43092dfdc12b2d242ade6bd1379fe1e7
Author: Bill Nottingham <notting@redhat.com>
Date:   Tue Oct 20 16:56:32 2009 -0400

    Load bonding driver if BONDING_OPTS is defined. (#516569)

 sysconfig/network-scripts/network-functions |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

commit f08000519addd0b13fada0eba4afef72871613c1
Author: Bill Nottingham <notting@redhat.com>
Date:   Thu Aug 27 13:25:41 2009 -0400

    Restore context of recreated files. (#519748)

 rc.d/rc.sysinit |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

commit 209e5e18cbd1cbb76f7e49c83854763c4e9f1d1c
Author: Bill Nottingham <notting@redhat.com>
Date:   Fri Oct 9 16:18:04 2009 -0400

    Do not try to activate subsets and layer raid and handle 'no raid disks'.
    
    - 'no raid disks' handling: (#527726, <mschmidt@redhat.com>)
    - no subsets: https://bugzilla.redhat.com/show_bug.cgi?id=524356

 rc.d/rc.sysinit |   15 +++++++--------
 1 files changed, 7 insertions(+), 8 deletions(-)

commit 40ddde4c4e6ad1dc855a8a3f4ae594d6c2d16179
Author: Bill Nottingham <notting@redhat.com>
Date:   Wed Oct 7 11:10:12 2009 -0400

    Add /var/lib/xend. (#526046)

 rwtab |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

commit 643c05c59a581663140e307f08732c6bef3d7622
Author: Bill Nottingham <notting@redhat.com>

src/initlog.c
/*
 * Copyright (c) 1999-2003 Red Hat, Inc. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA
 *
 */

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <libintl.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#define SYSLOG_NAMES
#include <syslog.h>

#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/un.h>
#include <sys/wait.h>

#define _(String) gettext((String))

#include <popt.h>

#include <regex.h>

#include "initlog.h"
#include "process.h"

static int logfacility=LOG_DAEMON;
static int logpriority=LOG_NOTICE;
static int reexec=0;
static int quiet=0;
int debug=0;

regex_t  **regList = NULL;

static int logEntries = 0;
struct logInfo *logData = NULL;

void readConfiguration(char *fname) {
    int fd,num=0;
    struct stat sbuf;
    char *data,*line, *d;
    regex_t *regexp;
    int lfac=-1,lpri=-1;
    
    if ((fd=open(fname,O_RDONLY))==-1) return;
    if (fstat(fd,&sbuf)) {
	    close(fd);
	    return;
    }
    d = data=malloc(sbuf.st_size+1);
    if (read(fd,data,sbuf.st_size)!=sbuf.st_size) {
	    close(fd);
	    free(data);
	    return;
    }
    close(fd);
    data[sbuf.st_size] = '\0';
    while ((line=getLine(&data))) {
	if (line[0]=='#') continue;
	if (!strncmp(line,"ignore ",7)) {
	    regexp = malloc(sizeof(regex_t));
	    if (!regcomp(regexp,line+7,REG_EXTENDED|REG_NOSUB|REG_NEWLINE)) {
		regList = realloc(regList,(num+2) * sizeof(regex_t *));
		regList[num] = regexp;
		regList[num+1] = NULL;
		num++;
	    }
	}
	if (!strncmp(line,"facility ",9)) {
	    lfac=atoi(line+9);
	    if ((lfac == 0) && strcmp(line+9,"0")) {
		int x =0;
		
		lfac = LOG_DAEMON;
		for (x=0;facilitynames[x].c_name;x++) {
		    if (!strcmp(line+9,facilitynames[x].c_name)) {
			lfac = facilitynames[x].c_val;
			break;
		    }
		}
	    }
	}
	if (!strncmp(line,"priority ",9)) {
	    lpri = atoi(line+9);
	    if ((lpri == 0) && strcmp(line+9,"0")) {
		int x=0;
		
		lpri = LOG_NOTICE;
		for (x=0;prioritynames[x].c_name;x++) {
		    if (!strcmp(line+9,prioritynames[x].c_name)) {
			lpri = prioritynames[x].c_val;
			break;
		    }
		}
	    }
	}
    }
    if (lfac!=-1) logfacility=lfac;
    if (lpri!=-1) logpriority=lpri;
    free(d);
}
    
char *getLine(char **data) {
    /* Get one line from data */
    /* Anything up to a carraige return (\r) or a backspace (\b) is discarded. */
    /* If this really bothers you, mail me and I might make it configurable. */
    /* It's here to avoid confilcts with fsck's progress bar. */
    char *x, *y;
    
    if (!*data) return NULL;
    x=*data;
    while (*x && (*x != '\n')) {
	while (*x && (*x != '\n') && (*x != '\r') && (*x != '\b')) x++;
	if (*x && (*x=='\r' || *x =='\b')) {
		*data = x+1;
		x++;
	}
    }
    if (*x) {
	x++;
    } else {
	if (x-*data) {
	    y=malloc(x-*data+1);
	    y[x-*data] = 0;
	    y[x-*data-1] = '\n';
	    memcpy(y,*data,x-*data);
	} else {
	    y=NULL;
	}
	*data = NULL;
	return y;
    }
    y = malloc(x-*data);
    y[x-*data-1] = 0;
    memcpy(y,*data,x-*data-1);
    *data = x;
    return y;
}

char **toArray(char *line, int *num) {
    /* Converts a long string into an array of lines. */
    char **lines;
    char *tmpline;
    
    *num = 0;
    lines = NULL;
    
    while ((tmpline=getLine(&line))) {
	if (!*num)
	  lines = (char **) malloc(sizeof(char *));
	else
	  lines = (char **) realloc(lines, (*num+1)*sizeof(char *));
	lines[*num] = tmpline;
	(*num)++;
    }
    return lines;
}

int trySocket() {
	int s;
	struct sockaddr_un addr;
 	
	s = socket(AF_LOCAL, SOCK_DGRAM, 0);
	if (s<0)
	  return 1;
   
	bzero(&addr,sizeof(addr));
	addr.sun_family = AF_LOCAL;
	strncpy(addr.sun_path,_PATH_LOG,sizeof(addr.sun_path)-1);

	if (connect(s,(struct sockaddr *) &addr,sizeof(addr))<0) {
		if (errno == EPROTOTYPE || errno == ECONNREFUSED) {
			DDEBUG("connect failed (EPROTOTYPE), trying stream\n");
			close(s);
			s = socket(AF_LOCAL, SOCK_STREAM, 0);
			if (connect(s,(struct sockaddr *) &addr, sizeof(addr)) < 0) {
				DDEBUG("connect failed: %s\n",strerror(errno));
				close(s);
				return 1;
			} 
			close(s);
			return 0;
		}
		close(s);
		DDEBUG("connect failed: %s\n",strerror(errno));
		return 1;
	} else {
		close(s);
		return 0;
	}
}

int logLine(struct logInfo *logEnt) {
    /* Logs a line... somewhere. */
    int x;
    struct stat statbuf;
    
    /* Don't log empty or null lines */
    if (!logEnt->line || !strcmp(logEnt->line,"\n")) return 0;
    
	
    if  ((stat(_PATH_LOG,&statbuf)==-1) || trySocket()) {
	DDEBUG("starting daemon failed, pooling entry %d\n",logEntries);
	logData=realloc(logData,(logEntries+1)*sizeof(struct logInfo));
	logData[logEntries].fac = logEnt->fac;
	logData[logEntries].pri = logEnt->pri;
	logData[logEntries].cmd = strdup(logEnt->cmd);
	logData[logEntries].line = strdup(logEnt->line);
	logEntries++;
    } else {
	if (logEntries>0) {
	    for (x=0;x<logEntries;x++) {
		DDEBUG("flushing log entry %d =%s=\n",x,logData[x].line);
		openlog(logData[x].cmd,0,logData[x].fac);
		syslog(logData[x].pri,"%s",logData[x].line);
		closelog();
	    }
	    free(logData);
	    logEntries = 0;
	}
	DDEBUG("logging =%s= via syslog\n",logEnt->line);
	openlog(logEnt->cmd,0,logEnt->fac);
	syslog(logEnt->pri,"%s",logEnt->line);
	closelog();
    }
    return 0;
}

int logEvent(char *cmd, int eventtype,char *string) {
    char *eventtable [] = {
	_("%s babbles incoherently"),
	_("%s succeeded"),
	_("%s failed"),
	_("%s cancelled at user request"),
	_("%s failed due to a failed dependency"),
	/* insert more here */
	NULL
    };
    int x=0,len, rc;
    struct logInfo logentry;
    
    if (cmd) {
	logentry.cmd = basename(cmd);
	if ((logentry.cmd[0] =='K' || logentry.cmd[0] == 'S') &&
	    ( logentry.cmd[1] >= '0' && logentry.cmd[1] <= '9' ) &&
	    ( logentry.cmd[2] >= '0' && logentry.cmd[2] <= '9' ) )
	  logentry.cmd+=3;
	logentry.cmd = strdup(logentry.cmd);
    } else
      logentry.cmd = strdup(_("(none)"));
    if (!string) {
      string = alloca(strlen(cmd)+1);
      strcpy(string,cmd);
    }
    
    while (eventtable[x] && x<eventtype) x++;
    if (!(eventtable[x])) x=0;
    
    len=strlen(eventtable[x])+strlen(string);
    logentry.line=malloc(len);
    snprintf(logentry.line,len,eventtable[x],string);
    
    logentry.pri = logpriority;
    logentry.fac = logfacility;
    
    rc = logLine(&logentry);
    free(logentry.line);
    free(logentry.cmd);
    return rc;
}

int logString(char *cmd, char *string) {
    struct logInfo logentry;
    int rc;
    
    if (cmd) {
	logentry.cmd = basename(cmd);
	if ((logentry.cmd[0] =='K' || logentry.cmd[0] == 'S') && 
	    ( logentry.cmd[1] >= '0' && logentry.cmd[1] <= 0x39 ) &&
	    ( logentry.cmd[2] >= '0' && logentry.cmd[2] <= 0x39 ) )
	  logentry.cmd+=3;
	logentry.cmd = strdup(logentry.cmd);
    } else
      logentry.cmd = strdup(_(""));
    logentry.line = strdup(string);
    logentry.pri = logpriority;
    logentry.fac = logfacility;
    
    rc = logLine(&logentry);
    free(logentry.line);
    free(logentry.cmd);
    return rc;
}

int processArgs(int argc, char **argv, int silent) {
    char *cmdname=NULL;
    char *conffile=NULL;
    int cmdevent=0;
    char *cmd=NULL;
    char *logstring=NULL;
    char *fac=NULL,*pri=NULL;
    int lfac=-1, lpri=-1;
    poptContext context;
    int rc;
    struct poptOption optTable[] = {
	POPT_AUTOHELP
	{ "conf", 0, POPT_ARG_STRING, &conffile, 0,
	  "configuration file (default: /etc/initlog.conf)", NULL
	},
	{ "name", 'n', POPT_ARG_STRING, &cmdname, 0,
	  "name of service being logged", NULL 
	},
	{ "event", 'e', POPT_ARG_INT, &cmdevent, 0,
	  "event being logged (see man page)", NULL
	},
	{ "cmd", 'c', POPT_ARG_STRING, &cmd, 0,
	  "command to run, logging output", NULL
	},
        { "debug", 'd', POPT_ARG_NONE, &debug, 0,
	  "print lots of verbose debugging info", NULL
	},
	{ "run", 'r', POPT_ARG_STRING, &cmd, 3,
	  "command to run, accepting input on open fd", NULL
	},
	{ "string", 's', POPT_ARG_STRING, &logstring, 0,
	  "string to log", NULL
	},
	{ "facility", 'f', POPT_ARG_STRING, &fac, 1,
	  "facility to log at (default: 'local7')", NULL
	},
	{ "priority", 'p', POPT_ARG_STRING, &pri, 2,
	  "priority to log at (default: 'notice')", NULL
	},
        { "quiet", 'q', POPT_ARG_NONE, &quiet, 0,
	  "suppress stdout/stderr", NULL
	},
        { 0, 0, 0, 0, 0, 0 }
    };
    
    context = poptGetContext("initlog", argc, argv, optTable, 0);
    
    while ((rc = poptGetNextOpt(context)) > 0) {
	switch (rc) {
	 case 1:
	    lfac=atoi(fac);
	    if ((lfac == 0) && strcmp(fac,"0")) {
		int x =0;
		
		lfac = LOG_DAEMON;
		for (x=0;facilitynames[x].c_name;x++) {
		    if (!strcmp(fac,facilitynames[x].c_name)) {
			lfac = facilitynames[x].c_val;
			break;
		    }
		}
	    }
	    break;
	 case 2:
	    lpri = atoi(pri);
	    if ((lpri == 0) && strcmp(pri,"0")) {
		int x=0;
		
		lpri = LOG_NOTICE;
		for (x=0;prioritynames[x].c_name;x++) {
		    if (!strcmp(pri,prioritynames[x].c_name)) {
			lpri = prioritynames[x].c_val;
			break;
		    }
		}
	    }
	    break;
	 case 3:
	    reexec = 1;
	    break;
	 default:
	    break;
	}
    }
      
    if ((rc < -1)) {
       if (!silent)
	 fprintf(stderr, "%s: %s\n",
		poptBadOption(context, POPT_BADOPTION_NOALIAS),
		poptStrerror(rc));
       
	return -1;
    }
    if ( (cmd && logstring) || (cmd && cmdname) ) {
        if (!silent)
	 fprintf(stderr, _("--cmd and --run are incompatible with --string or --name\n"));
	return -1;
    }
    if ( cmdname && (!logstring && !cmdevent)) {
        if (!silent)
	 fprintf(stderr, _("--name requires one of --event or --string\n"));
	return -1;
    }
    if (cmdevent && cmd) {
	    if (!silent)
	      fprintf(stderr, _("--cmd and --run are incompatible with --event\n"));
	    return -1;
    }
    if (conffile) {
	readConfiguration(conffile);
    } else {
	readConfiguration("/etc/initlog.conf");
    }
    if (cmd) {
	    while (isspace(*cmd)) cmd++;
    }
    if (lpri!=-1) logpriority=lpri;
    if (lfac!=-1) logfacility=lfac;
    if (cmdevent) {
	logEvent(cmdname,cmdevent,logstring);
    } else if (logstring) {
	logString(cmdname,logstring);
    } else if ( cmd && *cmd) {
	return(runCommand(cmd,reexec,quiet,debug));
    } else {
        if (!silent)
	 fprintf(stderr,"nothing to do!\n");
	return -1;
    }
   return 0;
}

int main(int argc, char **argv) {

    setlocale(LC_ALL,"");
    bindtextdomain("initlog","/etc/locale");
    textdomain("initlog");
    fprintf(stderr, _("WARNING: initlog is deprecated and will be removed in a future release\n"));
    exit(processArgs(argc,argv,0));
}
